File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
# 4칸 내려가 -> 3칸 오른쪽 -> 2칸 대간선 왼 -> 1칸 내려가
# 5칸 내려가 -> 4칸 오른쪽 -> 3칸 대각선 위 -> 2칸 내려가 -> 1칸 오른쪽
# 왜안올라가
def solution(n):
    answer = []
    
    max_value = n * (n+1) // 2
    snail = [[0] * i for i in range(1, n + 1)]
    direction = ["down", "right", "up"]
    d = 0
    
    num = 1
    x, y = 0, 0
    change = n
    while num <= max_value:
        
        for i in range(change):
            # print('x, y = ', x, y)
            snail[x][y] = num
            # print('snail[x][y] = ', snail[x][y])
            num += 1
            if i < change - 1:
                if direction[d] == "down":
                    x += 1
                elif direction[d] == "right":
                    y += 1
                else:
                    x -= 1
                    y -= 1
                    
        d = (d + 1) % 3
        
        if direction[d] == "down":
            x += 1
        elif direction[d] == "right":
            y += 1
        elif direction[d] == "up":
            x -= 1
            y -= 1
        # print("change")
        change -= 1
        
    # print(snail)
    
    for number in snail:
        answer.extend(number)
    
    return answer

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
class Solution {
    public int[] solution(int n) {
        int last = n * n - n * (n - 1) / 2;
        int[][] triangle = new int[n][];
        for (int i = 0; i < n ; i++)
            triangle[i] = new int[i+1];
        int[][] direction = {
            {1, 0},
            {0, 1},
            {-1, -1}
        };
        
        int x = 0;
        int y = 0;
        int dir = 0;
        for (int i = 1; i <= last; i++) {
            triangle[x][y] = i;
            int nx = x + direction[dir][0];
            int ny = y + direction[dir][1];
            if (ny < 0 || ny > x || nx < 0 || nx >= n || triangle[nx][ny] != 0) {
                dir = (dir + 1) % 3;
                nx = x + direction[dir][0];
                ny = y + direction[dir][1];
            }
            x = nx;
            y = ny;
        }
        
        int[] answer = new int[last];
        int idx = 0;
        for (int i = 0; i < n; i++) {
            int[] rowNum = triangle[i];
            for (int num : rowNum) {
                answer[idx++] = num;
            }
        }
        return answer;
    }
}